# Segmenting and Clustering Neighborhood in Toronto

### Part 1: Prepare dataframe fetching data from webpage table

In [36]:
# import libraries
import requests
import lxml.html as lh
import pandas as pd

In [37]:
# setting url for the source data
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
# Create a handle, page, to handle the contents of the website
page = requests.get(url)
# Store the contents of the website under doc
doc = lh.fromstring(page.content)
# Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

##### For sanity check, ensure that all the rows have the same width. If not, we probably got something more than just the table.

In [38]:
# Check the length of the first 12 rows (Sanity Check)
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

In [39]:
# Create empty list
col=[]
i=0
# For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print('%d:"%s"' %(i,name))
    col.append((name,[]))

1:"Postal Code
"
2:"Borough
"
3:"Neighbourhood
"


In [40]:
# Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T  is our j'th row
    T=tr_elements[j]
    
    # If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    # i is the index of our column
    i=0
    
    # Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        # Check if row is empty
        if i>0:
        # Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        # Append the data to the empty list of the i'th column
        col[i][1].append(data)
        # Increment i for the next column
        i+=1

##### Just to be sure, let’s check the length of each column. Ideally, they should all be the same.

In [41]:
[len(C) for (title,C) in col]

[181, 181, 181]

##### Create the DataFrame

In [42]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

##### Looking at the top 5 cells on the DataFrame

In [43]:
df.head()

Postal Code\n           Borough\n              Neighbourhood\n
0         M1A\n      Not assigned\n               Not assigned\n
1         M2A\n      Not assigned\n               Not assigned\n
2         M3A\n        North York\n                  Parkwoods\n
3         M4A\n        North York\n           Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park, Harbourfront\n

##### Renaming column headings to remove \n and to phrase them as needed

In [44]:
df.rename(columns={"Postal Code\n": "Postal Code", "Borough\n": "Borough", "Neighbourhood\n": "Neighborhood"}, inplace=True)


In [45]:
df.head()

Postal Code             Borough                 Neighborhood
0       M1A\n      Not assigned\n               Not assigned\n
1       M2A\n      Not assigned\n               Not assigned\n
2       M3A\n        North York\n                  Parkwoods\n
3       M4A\n        North York\n           Victoria Village\n
4       M5A\n  Downtown Toronto\n  Regent Park, Harbourfront\n

##### Ignore cells with a borough that is 'Not assigned'

In [46]:
df = df[~df['Borough'].str.contains('Not assigned')]

In [47]:
df

Postal Code                  Borough  \
2         M3A\n             North York\n   
3         M4A\n             North York\n   
4         M5A\n       Downtown Toronto\n   
5         M6A\n             North York\n   
6         M7A\n       Downtown Toronto\n   
8         M9A\n              Etobicoke\n   
9         M1B\n            Scarborough\n   
11        M3B\n             North York\n   
12        M4B\n              East York\n   
13        M5B\n       Downtown Toronto\n   
14        M6B\n             North York\n   
17        M9B\n              Etobicoke\n   
18        M1C\n            Scarborough\n   
20        M3C\n             North York\n   
21        M4C\n              East York\n   
22        M5C\n       Downtown Toronto\n   
23        M6C\n                   York\n   
26        M9C\n              Etobicoke\n   
27        M1E\n            Scarborough\n   
30        M4E\n           East Toronto\n   
31        M5E\n       Downtown Toronto\n   
32        M6E\n                   York\n   
36        M1G\n            Scarborough\n   
39        M4G\n              East York\n   
40        M5G\n       Downtown Toronto\n   
41        M6G\n       Downtown Toronto\n   
45        M1H\n            Scarborough\n   
46        M2H\n             North York\n   
47        M3H\n             North York\n   
48        M4H\n              East York\n   
49        M5H\n       Downtown Toronto\n   
50        M6H\n           West Toronto\n   
54        M1J\n            Scarborough\n   
55        M2J\n             North York\n   
56        M3J\n             North York\n   
57        M4J\n              East York\n   
58        M5J\n       Downtown Toronto\n   
59        M6J\n           West Toronto\n   
63        M1K\n            Scarborough\n   
64        M2K\n             North York\n   
65        M3K\n             North York\n   
66        M4K\n           East Toronto\n   
67        M5K\n       Downtown Toronto\n   
68        M6K\n           West Toronto\n   
72        M1L\n            Scarborough\n   
73        M2L\n             North York\n   
74        M3L\n             North York\n   
75        M4L\n           East Toronto\n   
76        M5L\n       Downtown Toronto\n   
77        M6L\n             North York\n   
80        M9L\n             North York\n   
81        M1M\n            Scarborough\n   
82        M2M\n             North York\n   
83        M3M\n             North York\n   
84        M4M\n           East Toronto\n   
85        M5M\n             North York\n   
86        M6M\n                   York\n   
89        M9M\n             North York\n   
90        M1N\n            Scarborough\n   
91        M2N\n             North York\n   
92        M3N\n             North York\n   
93        M4N\n        Central Toronto\n   
94        M5N\n        Central Toronto\n   
95        M6N\n                   York\n   
98        M9N\n                   York\n   
99        M1P\n            Scarborough\n   
100       M2P\n             North York\n   
102       M4P\n        Central Toronto\n   
103       M5P\n        Central Toronto\n   
104       M6P\n           West Toronto\n   
107       M9P\n              Etobicoke\n   
108       M1R\n            Scarborough\n   
109       M2R\n             North York\n   
111       M4R\n        Central Toronto\n   
112       M5R\n        Central Toronto\n   
113       M6R\n           West Toronto\n   
114       M7R\n            Mississauga\n   
116       M9R\n              Etobicoke\n   
117       M1S\n            Scarborough\n   
120       M4S\n        Central Toronto\n   
121       M5S\n       Downtown Toronto\n   
122       M6S\n           West Toronto\n   
126       M1T\n            Scarborough\n   
129       M4T\n        Central Toronto\n   
130       M5T\n       Downtown Toronto\n   
135       M1V\n            Scarborough\n   
138       M4V\n        Central Toronto\n   
139       M5V\n       Downtown Toronto\n   
142       M8V\n              Etobicoke\n   
143       M9V\n              Etobicoke\n   


##### Remove last undesired row

In [48]:
df = df[:-1]

In [49]:
df.tail()

Postal Code             Borough  \
160       M8X\n         Etobicoke\n   
165       M4Y\n  Downtown Toronto\n   
168       M7Y\n      East Toronto\n   
169       M8Y\n         Etobicoke\n   
178       M8Z\n         Etobicoke\n   

                                          Neighborhood  
160    The Kingsway, Montgomery Road, Old Mill North\n  
165                             Church and Wellesley\n  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...

##### Remove \n from all rows

In [50]:
df_final = df.replace("\n", "", regex=True)

In [51]:
df_final.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

##### Restting the index to modified dataframe

In [52]:
df_final.reset_index(inplace=True, drop=True)

In [53]:
df_final.head(10)

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

##### Checking shape and value counts to ensure no duplicates of Postal Codes

In [90]:
df_final['Postal Code'].value_counts()

M4K    1
M4M    1
M8Y    1
M6P    1
M5C    1
M4S    1
M6A    1
M1H    1
M2N    1
M5V    1
M5T    1
M6C    1
M5X    1
M8V    1
M9C    1
M1W    1
M4X    1
M6J    1
M2L    1
M1K    1
M5L    1
M2J    1
M4W    1
M4E    1
M6H    1
M3L    1
M2K    1
M9A    1
M3H    1
M5E    1
M9L    1
M6N    1
M6E    1
M3B    1
M9M    1
M4J    1
M3K    1
M2M    1
M4G    1
M5P    1
M4V    1
M7A    1
M1B    1
M1G    1
M3N    1
M9B    1
M1T    1
M4Y    1
M4L    1
M7R    1
M9W    1
M6L    1
M5R    1
M3J    1
M5J    1
M4H    1
M5H    1
M4N    1
M1J    1
M1E    1
M8W    1
M4B    1
M5K    1
M5A    1
M6S    1
M1S    1
M4P    1
M5G    1
M1X    1
M3C    1
M1M    1
M3A    1
M5B    1
M1C    1
M9R    1
M8Z    1
M4C    1
M9N    1
M1L    1
M9V    1
M7Y    1
M5M    1
M9P    1
M2R    1
M1V    1
M8X    1
M1N    1
M6B    1
M4R    1
M6K    1
M5N    1
M5W    1
M2P    1
M6M    1
M5S    1
M2H    1
M4A    1
M1P    1
M6G    1
M4T    1
M3M    1
M1R    1
M6R    1
Name: Postal Code, dtype: int64

In [55]:
df_final.shape

(103, 3)

#### Since Value Counts are equal to number of rows (103), there are no duplicate Postal Codes, for which 'Neighborhood' needs to be combined.

### ----- End of Part-1 -----

### Part -2 : Create Dataframe of Geospacial Data

In [56]:
import pandas as pd

In [57]:
# Create a dataframe from csv file
df_gsc = pd.read_csv("Geospatial_Coordinates.csv")

In [58]:
# display top 5 rows
df_gsc.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [59]:
# dimensions of the dataframe
df_gsc.shape

(103, 3)

In [60]:
# set common index on both the dataframes
df_final.set_index('Postal Code')

Borough  \
Postal Code                     
M3A                North York   
M4A                North York   
M5A          Downtown Toronto   
M6A                North York   
M7A          Downtown Toronto   
M9A                 Etobicoke   
M1B               Scarborough   
M3B                North York   
M4B                 East York   
M5B          Downtown Toronto   
M6B                North York   
M9B                 Etobicoke   
M1C               Scarborough   
M3C                North York   
M4C                 East York   
M5C          Downtown Toronto   
M6C                      York   
M9C                 Etobicoke   
M1E               Scarborough   
M4E              East Toronto   
M5E          Downtown Toronto   
M6E                      York   
M1G               Scarborough   
M4G                 East York   
M5G          Downtown Toronto   
M6G          Downtown Toronto   
M1H               Scarborough   
M2H                North York   
M3H                North York   
M4H                 East York   
M5H          Downtown Toronto   
M6H              West Toronto   
M1J               Scarborough   
M2J                North York   
M3J                North York   
M4J                 East York   
M5J          Downtown Toronto   
M6J              West Toronto   
M1K               Scarborough   
M2K                North York   
M3K                North York   
M4K              East Toronto   
M5K          Downtown Toronto   
M6K              West Toronto   
M1L               Scarborough   
M2L                North York   
M3L                North York   
M4L              East Toronto   
M5L          Downtown Toronto   
M6L                North York   
M9L                North York   
M1M               Scarborough   
M2M                North York   
M3M                North York   
M4M              East Toronto   
M5M                North York   
M6M                      York   
M9M                North York   
M1N               Scarborough   
M2N                North York   
M3N                North York   
M4N           Central Toronto   
M5N           Central Toronto   
M6N                      York   
M9N                      York   
M1P               Scarborough   
M2P                North York   
M4P           Central Toronto   
M5P           Central Toronto   
M6P              West Toronto   
M9P                 Etobicoke   
M1R               Scarborough   
M2R                North York   
M4R           Central Toronto   
M5R           Central Toronto   
M6R              West Toronto   
M7R               Mississauga   
M9R                 Etobicoke   
M1S               Scarborough   
M4S           Central Toronto   
M5S          Downtown Toronto   
M6S              West Toronto   
M1T               Scarborough   
M4T           Central Toronto   
M5T          Downtown Toronto   
M1V               Scarborough   
M4V           Central Toronto   
M5V          Downtown Toronto   
M8V                 Etobicoke   
M9V                 Etobicoke   
M1W               Scarborough   
M4W          Downtown Toronto   
M5W          Downtown Toronto   
M8W                 Etobicoke   
M9W                 Etobicoke   
M1X               Scarborough   
M4X          Downtown Toronto   
M5X          Downtown Toronto   
M8X                 Etobicoke   
M4Y          Downtown Toronto   
M7Y              East Toronto   
M8Y                 Etobicoke   
M8Z                 Etobicoke   

                                                  Neighborhood  
Postal Code                                                     
M3A                                                  Parkwoods  
M4A                                           Victoria Village  
M5A                                  Regent Park, Harbourfront  
M6A                           Lawrence Manor, Lawrence Heights  
M7A                Queen's Park, Ontario Provincial Government  
M9A                    Islington Avenue, Humber Valley Village  
M1B                                 

In [61]:
# set common index on both the dataframes
df_gsc.set_index('Postal Code')

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476
M1J          43.744734 -79.239476
M1K          43.727929 -79.262029
M1L          43.711112 -79.284577
M1M          43.716316 -79.239476
M1N          43.692657 -79.264848
M1P          43.757410 -79.273304
M1R          43.750072 -79.295849
M1S          43.794200 -79.262029
M1T          43.781638 -79.304302
M1V          43.815252 -79.284577
M1W          43.799525 -79.318389
M1X          43.836125 -79.205636
M2H          43.803762 -79.363452
M2J          43.778517 -79.346556
M2K          43.786947 -79.385975
M2L          43.757490 -79.374714
M2M          43.789053 -79.408493
M2N          43.770120 -79.408493
M2P          43.752758 -79.400049
M2R          43.782736 -79.442259
M3A          43.753259 -79.329656
M3B          43.745906 -79.352188
M3C          43.725900 -79.340923
M3H          43.754328 -79.442259
M3J          43.767980 -79.487262
M3K          43.737473 -79.464763
M3L          43.739015 -79.506944
M3M          43.728496 -79.495697
M3N          43.761631 -79.520999
M4A          43.725882 -79.315572
M4B          43.706397 -79.309937
M4C          43.695344 -79.318389
M4E          43.676357 -79.293031
M4G          43.709060 -79.363452
M4H          43.705369 -79.349372
M4J          43.685347 -79.338106
M4K          43.679557 -79.352188
M4L          43.668999 -79.315572
M4M          43.659526 -79.340923
M4N          43.728020 -79.388790
M4P          43.712751 -79.390197
M4R          43.715383 -79.405678
M4S          43.704324 -79.388790
M4T          43.689574 -79.383160
M4V          43.686412 -79.400049
M4W          43.679563 -79.377529
M4X          43.667967 -79.367675
M4Y          43.665860 -79.383160
M5A          43.654260 -79.360636
M5B          43.657162 -79.378937
M5C          43.651494 -79.375418
M5E          43.644771 -79.373306
M5G          43.657952 -79.387383
M5H          43.650571 -79.384568
M5J          43.640816 -79.381752
M5K          43.647177 -79.381576
M5L          43.648198 -79.379817
M5M          43.733283 -79.419750
M5N          43.711695 -79.416936
M5P          43.696948 -79.411307
M5R          43.672710 -79.405678
M5S          43.662696 -79.400049
M5T          43.653206 -79.400049
M5V          43.628947 -79.394420
M5W          43.646435 -79.374846
M5X          43.648429 -79.382280
M6A          43.718518 -79.464763
M6B          43.709577 -79.445073
M6C          43.693781 -79.428191
M6E          43.689026 -79.453512
M6G          43.669542 -79.422564
M6H          43.669005 -79.442259
M6J          43.647927 -79.419750
M6K          43.636847 -79.428191
M6L          43.713756 -79.490074
M6M          43.691116 -79.476013
M6N          43.673185 -79.487262
M6P          43.661608 -79.464763
M6R          43.648960 -79.456325
M6S          43.651571 -79.484450
M7A          43.662301 -79.389494
M7R          43.636966 -79.615819
M7Y          43.662744 -79.321558
M8V          43.605647 -79.501321
M8W          43.602414 -79.543484
M8X          43.653654 -79.506944
M8Y          43.636258 -79.498509
M8Z          43.628841 -79.520999
M9A          43.667856 -79.532242
M9B          43.650943 -79.554724
M9C          43.643515 -79.577201
M9L          43.756303 -79.565963
M9M          43.724766 -79.532242
M9N          43.706876 -79.518188
M9P          43.696319 -79.532242
M9R          43.688905 -79.554724
M9V          43.739416 -79.588437
M9W          43.706748 -79.594054

In [62]:
# merge both the dataframes in a third dataframe
df_merged = df_final.merge(df_gsc, on="Postal Code", left_index=True)

In [63]:
# display top 12 rows
df_merged.head(12)

Postal Code           Borough  \
25         M3A        North York   
34         M4A        North York   
53         M5A  Downtown Toronto   
71         M6A        North York   
85         M7A  Downtown Toronto   
93         M9A         Etobicoke   
0          M1B       Scarborough   
26         M3B        North York   
35         M4B         East York   
54         M5B  Downtown Toronto   
72         M6B        North York   
94         M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
25                                          Parkwoods  43.753259 -79.329656  
34                                   Victoria Village  43.725882 -79.315572  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
71                   Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
85        Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
93            Islington Avenue, Humber Valley Village  43.667856 -79.532242  
0                                      Malvern, Rouge  43.806686 -79.194353  
26                                          Don Mills  43.745906 -79.352188  
35                    Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
54                           Garden District, Ryerson  43.657162 -79.378937  
72                                          Glencairn  43.709577 -79.445073  
94  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

### ----- End of Part-2 -----

### Part-3: Exploring and clustering neighborhoods in Toronto

## Create a map of Toronto with neighborhoods superimposed on top.

### Work with only boroughs that contain the word "Toronto"

In [64]:
df_merged = df_merged[df_merged['Borough'].str.contains('Toronto')]

In [65]:
df_merged.head(20)

Postal Code           Borough  \
53         M5A  Downtown Toronto   
85         M7A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
37         M4E      East Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
75         M6G  Downtown Toronto   
58         M5H  Downtown Toronto   
76         M6H      West Toronto   
59         M5J  Downtown Toronto   
77         M6J      West Toronto   
41         M4K      East Toronto   
60         M5K  Downtown Toronto   
78         M6K      West Toronto   
42         M4L      East Toronto   
61         M5L  Downtown Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
63         M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
85        Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
37                                        The Beaches  43.676357 -79.293031  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
75                                           Christie  43.669542 -79.422564  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
76                       Dufferin, Dovercourt Village  43.669005 -79.442259  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
77                           Little Portugal, Trinity  43.647927 -79.419750  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
78       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
63                                           Roselawn  43.711695 -79.416936

In [66]:
df_merged.shape

(39, 5)

#### Folium and Geopy are downloaded separately from Anaconda cloud

In [67]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [68]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_merged['Borough'].unique()),
        df_merged.shape[0]
    )
)

The dataframe has 4 boroughs and 39 neighborhoods.


In [69]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [70]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_merged['Latitude'], df_merged['Longitude'], df_merged['Borough'], df_merged['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [71]:
df_merged.reset_index(inplace=True, drop=True)

In [72]:
df_merged.head(39)

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                 North Toronto West,  Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

In [73]:
df_merged.dtypes

Postal Code      object
Borough          object
Neighborhood     object
Latitude        float64
Longitude       float64
dtype: object

In [74]:
df_merged.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [75]:
neighborhood_latitude = df_merged.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_merged.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_merged.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


##### dropped 'Neighborhood' column just to reduce number of records to be processed, 'Postal Code' can be used instead.

In [76]:
df_merged_wo = df_merged.drop('Neighborhood', 1)

In [77]:
df_merged_wo.head()

Postal Code           Borough   Latitude  Longitude
0         M5A  Downtown Toronto  43.654260 -79.360636
1         M7A  Downtown Toronto  43.662301 -79.389494
2         M5B  Downtown Toronto  43.657162 -79.378937
3         M5C  Downtown Toronto  43.651494 -79.375418
4         M4E      East Toronto  43.676357 -79.293031

In [78]:
# one hot encoding
df_onehot = pd.get_dummies(df_merged, columns=["Borough", "Postal Code"], prefix="")

In [79]:
df_onehot.shape

(39, 46)

In [80]:
df_onehot.head()

Neighborhood   Latitude  Longitude  \
0                    Regent Park, Harbourfront  43.654260 -79.360636   
1  Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                     Garden District, Ryerson  43.657162 -79.378937   
3                               St. James Town  43.651494 -79.375418   
4                                  The Beaches  43.676357 -79.293031   

   _Central Toronto  _Downtown Toronto  _East Toronto  _West Toronto  _M4E  \
0                 0                  1              0              0     0   
1                 0                  1              0              0     0   
2                 0                  1              0              0     0   
3                 0                  1              0              0     0   
4                 0                  0              1              0     1   

   _M4K  _M4L  _M4M  _M4N  _M4P  _M4R  _M4S  _M4T  _M4V  _M4W  _M4X  _M4Y  \
0     0     0     0     0     0     0     0     0     0     0     0     0   
1     0     0     0     0     0     0     0     0     0     0     0     0   
2     0     0     0     0     0     0     0     0     0     0     0     0   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     0     0     0     0     0     0     0     0   

   _M5A  _M5B  _M5C  _M5E  _M5G  _M5H  _M5J  _M5K  _M5L  _M5N  _M5P  _M5R  \
0     1     0     0     0     0     0     0     0     0     0     0     0   
1     0     0     0     0     0     0     0     0     0     0     0     0   
2     0     1     0     0     0     0     0     0     0     0     0     0   
3     0     0     1     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     0     0     0     0     0     0     0     0   

   _M5S  _M5T  _M5V  _M5W  _M5X  _M6G  _M6H  _M6J  _M6K  _M6P  _M6R  _M6S  \
0     0     0     0     0     0     0     0     0     0     0     0     0   
1     0     0     0     0     0     0     0     0     0     0     0     0   
2     0     0     0     0     0     0     0     0     0     0     0     0   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     0     0     0     0     0     0     0     0   

   _M7A  _M7Y  
0     0     0  
1     1     0  
2     0     0  
3     0     0  
4     0     0

In [83]:
df_onehot.drop('Neighborhood', 1, inplace=True)

In [84]:
# set number of clusters
kclusters = 4


# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(df_onehot)
# k_means = KMeans(init="k-means++", n_clusters=5, n_init=12)
# k_means.fit(df_merged_clustering)
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]
# k_means_labels = k_means.labels_
# k_means_labels

array([0, 0, 0, 0, 2, 0, 0, 0, 0, 1])

In [85]:
# add clustering labels
df_onehot.insert(0, 'Cluster Labels', kmeans.labels_)

In [86]:
df_onehot.head()

Cluster Labels   Latitude  Longitude  _Central Toronto  _Downtown Toronto  \
0               0  43.654260 -79.360636                 0                  1   
1               0  43.662301 -79.389494                 0                  1   
2               0  43.657162 -79.378937                 0                  1   
3               0  43.651494 -79.375418                 0                  1   
4               2  43.676357 -79.293031                 0                  0   

   _East Toronto  _West Toronto  _M4E  _M4K  _M4L  _M4M  _M4N  _M4P  _M4R  \
0              0              0     0     0     0     0     0     0     0   
1              0              0     0     0     0     0     0     0     0   
2              0              0     0     0     0     0     0     0     0   
3              0              0     0     0     0     0     0     0     0   
4              1              0     1     0     0     0     0     0     0   

   _M4S  _M4T  _M4V  _M4W  _M4X  _M4Y  _M5A  _M5B  _M5C  _M5E  _M5G  _M5H  \
0     0     0     0     0     0     0     1     0     0     0     0     0   
1     0     0     0     0     0     0     0     0     0     0     0     0   
2     0     0     0     0     0     0     0     1     0     0     0     0   
3     0     0     0     0     0     0     0     0     1     0     0     0   
4     0     0     0     0     0     0     0     0     0     0     0     0   

   _M5J  _M5K  _M5L  _M5N  _M5P  _M5R  _M5S  _M5T  _M5V  _M5W  _M5X  _M6G  \
0     0     0     0     0     0     0     0     0     0     0     0     0   
1     0     0     0     0     0     0     0     0     0     0     0     0   
2     0     0     0     0     0     0     0     0     0     0     0     0   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     0     0     0     0     0     0     0     0   

   _M6H  _M6J  _M6K  _M6P  _M6R  _M6S  _M7A  _M7Y  
0     0     0     0     0     0     0     0     0  
1     0     0     0     0     0     0     1     0  
2     0     0     0     0     0     0     0     0  
3     0     0     0     0     0     0     0     0  
4     0     0     0     0     0     0     0     0

In [87]:
df_onehot.shape

(39, 46)

In [88]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, cluster in zip(df_onehot['Latitude'], df_onehot['Longitude'], df_onehot['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### ---- End of Part-3 -----